### PR TITLE
feat: use uid in K8s deployment and docker user

### PR DESCRIPTION
**Pull request by saidsef**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- ~[ ] Breaking change (fix or feature that would cause existing functionality to change)~

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](../CONTRIBUTING.md) guide
- [x] I have added CI tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by saidsef — COMMENTED** [review]
:robot: 
modified: deployment/k8s/base/server/deployment.yml (adding security settings for K8s unprivileged deployment)
- added runAsGroup, runAsNonRoot, runAsUser with value 10001
- dropped all capabilities except those needed

modified: deployment/k8s/base/ui/deployment.yml (applying same changes as server)

modified: function/Dockerfile (using non-root user in Docker)
- added chown to owner 10001 all inside /opt/tika
- changed Docker user to 10001

modified: function/Dockerfile.server (applying same changes as main function Dockerfile)

modified: ui/Dockerfile (applying same changes as main function Dockerfile)

Applying UID=10001 is a good security practice as it ensures that apps are run with unique non-root users, reducing damage if one app is compromised. Plus, this prevents the potential permission clashing with the host filesystem where this container is deployed (K8s provides an option for running containers as arbitrary user IDs).

Generally there are no suggestions or issues, except that ensuring PATH considerations are correct for tesseract, Tika libraries, and other executables. Please ensure the ENV variables and dependencies are properly adjusted for alpine based images. 

Summary of the review is to ensure security enhancements by applying unique UID to Docker processes and considering overall best security practices. They will not affect k8s deployments unless k8s is also configured accordingly. Verify the required dependencies and paths are configured correctly for the alpine-based Docker images. 

<sub><sub>> HuggingFaceH4/starchat2-15b-v0.1 [pr | 0]</sub></sub>

**Review by saidsef — COMMENTED** [review]
:robot: 
Hello, Said Sef!

Here's a summary of my evaluation regarding your PR: 

**Summary:** The proposed change addresses a security concern by setting a non-root user (uid 10001) in the Docker images for the server, UI, and function. It provides superior security by minimizing the attack surface of these applications and enforcing file system immutability.

**Improvement suggestions:**

I'm glad to see that you added changes to Dockerfiles. As per best practices, it's a fantastic idea to run Docker containers with a non-root user to enhance security and limit potential damage if a container gets compromised. 

However, I would like you to pay attention to certain lines in your patch:

1. deployment/k8s/base/server/deployment.yml
    - Gender-neutral language is used, "privileged: false" is correctly formatted.

2. deployment/k8s/base/ui/deployment.yml
   - The Dockerfile path is using Unix-style line endings (LF), and indentation is consistent.

3. function/Dockerfile
    - Efficiency: The Dockerfile sets the ownership of the `/opt/tika` directory only once at the end, which prevents redundancy and clutter.

4. function/Dockerfile.server
    - Syntax is correct; alignment and spacing follow standard conventions.

5. ui/Dockerfile
    - The Dockerfile charges the `chown` command correctly to the `/home/node` directory, not the root directory (`.`).

As per clear formatting practices, I have noticed small discrepancies in your patch. Please make sure the file paths and maintained consistency in indention, as this increases overall readability.

**In conclusion,** your PR is well-written and addresses an essential issue for enhancing the security of your microservices infrastructure. Keep up the good work! Let me know if you require further assistance or clarification.

Please consider revising and reformatting the file paths and maintain consistent indentation to ensure your PR is clear and readable. You may also consider adding comments for even greater clarity. Once those revisions are made, I'll be delighted to provide additional feedback.

Thank you for contributing to the better security of our microservices environment!

Best regards,
Reviewer Name (British English) 

<sub><sub>> HuggingFaceH4/starchat2-15b-v0.1 [pr | 1]</sub></sub>

**Review by saidsef — COMMENTED** [review]
:robot: 
### Summary of New Suggestions for Enhancements

The pull request introduces several enhancements by using a specific UID (10001) for Kubernetes deployments and Docker user settings across various Dockerfiles. Here are the key changes:

1. **Kubernetes Deployments:**
   - In `deployment/k8s/base/server/deployment.yml` and `deployment/k8s/base/ui/deployment.yml`, the `runAsUser`, `runAsGroup`, and `runAsNonRoot` settings are added to ensure non-root execution.
     ```yaml
     spec:
       containers:
         ...
         securityContext:
           runAsUser: 10001
           runAsGroup: 10001
           runAsNonRoot: true
     ```

2. **Docker User Settings:**
   - In `function/Dockerfile`, `function/Dockerfile.server`, and `ui/Dockerfile`, the user is changed from `nobody` to `10001` and the ownership of directories is updated accordingly.
     ```dockerfile
     USER 10001

     # Example of directory ownership change
     chown 10001:10001 /opt/tika
     ```

### Comments on Issues

- **Consistency:** The use of a specific UID (10001) across all Dockerfiles and Kubernetes deployments ensures consistency in user and group settings.
- **Security:** Running as a non-root user reduces the risk of privilege escalation attacks.
- **Best Practices:** Using a consistent UID aligns with best practices for container security and management.

### No Issues Identified

No significant issues were found with the code changes. The modifications appear to be well-intentioned and align with standard practices for securing container environments. 

<sub><sub>> llama-3.1-sonar-small-128k-chat [pr | 0]</sub></sub>
